### PR TITLE
Fix OpenMP-related segfault issue and additional updates in CRTM modules

### LINKED ISSUE #19: memory / NaN issues and underflow in v3
`develop` branch, tested with gfortran 12.0 on macos. 

```
The following tests FAILED:
	 36 - test_k_matrix_Simple_atms_n21 (SEGFAULT)
	 37 - test_k_matrix_Simple_cris-fsr_n21 (SEGFAULT)
	 38 - test_k_matrix_Simple_v.abi_g18 (SEGFAULT)
	 39 - test_k_matrix_Simple_modis_aqua (SEGFAULT)
	 67 - test_adjoint_Simple_atms_n21 (Failed)
	 68 - test_adjoint_Simple_cris-fsr_n21 (Failed)
	 69 - test_adjoint_Simple_v.abi_g18 (Failed)
	 70 - test_adjoint_Simple_modis_aqua (Failed)
```

```
36: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
36: 
<...>
1/1 Test #36: test_k_matrix_Simple_atms_n21 ....***Exception: SegFault  0.28 sec
```

```
67:      Comparing calculated results with saved ones...
67:  test_Simple(INFORMATION) : Atmosphere_AD save file does not exist. Creating...
67:  test_Simple(INFORMATION) : Surface_AD save file does not exist. Creating...
67:  test_Simple(FAILURE) : Atmosphere_AD Adjoints are different!
67: STOP 1
1/1 Test #67: test_adjoint_Simple_atms_n21 .....***Failed    0.17 sec

0% tests passed, 1 tests failed out of 1
```

This one is more odd, beause it creates the result, and compares it immediately, and fails that comparison.  so there's potentially an issue in the comparison structures or in the structures being passed.  The forward model and TL model seems to work fine.  
My first suspicion is that the aerosol structure is working properly.

Looking a bit deeper, the forward operator for modis_aqua is outputting NaNs in the aerosol portion of the atmosphere structure.  I @chengdang  I suspect that the something might be mismatched in the aerosol code update -- do the ctests need to be updated?

### LINKED ISSUE #29: test_adjoint_Simple
With the CRTM v3 test branch ([feature/v241_merge_test](https://github.com/JCSDA/CRTMv3/tree/feature/v241_merge_test)), the other group of tests constantly failed are `test_adjoint_Simple`, see issue #19 :

```
The following tests FAILED:
	121 - test_adjoint_Simple_atms_n21 (Failed)
	122 - test_adjoint_Simple_cris-fsr_n21 (Failed)
	123 - test_adjoint_Simple_v.abi_g18 (Failed)
	124 - test_adjoint_Simple_atms_npp (Failed)
	125 - test_adjoint_Simple_cris399_npp (Failed)
	126 - test_adjoint_Simple_v.abi_gr (Failed)
	127 - test_adjoint_Simple_abi_g18 (Failed)
	128 - test_adjoint_Simple_modis_aqua (Failed)
```

### LINKED ISSUE #33: tracking failures in check_crtm.F90
Noticed that check_crtm.F90 is crashing in v3, when it reaches the point in CRTM_Forward checking if the option Apply_NLTE_Correction is true.   Somewhere in the code, the Opt% structure is being corrupted. 

Also, the profile_solution function is not returning properly, when the function is called, all code after the function are not executed.  Probably related to the above issue.

### PR DESCRIPTION
**Pull request by BenjaminTJohnson**
## Description
This pull request addresses several updates and fixes in the CRTM module, mainly a recurring OpenMP-related segmentation fault issue.

Here is the summary of the changes:

- **OpenMP-related Segfault Issue:**
  The main highlight of this pull request is the introduction of the `Opt` options parameter into the `Post_Process_RTSolution` method in multiple CRTM modules, which effectively resolves an OpenMP-related segfault issue.
  
  This fix is introduced in the following modules:
  - `CRTM_Adjoint_Module.f90` (commit `c6da462216311df5905c554dbcb60b27685dc3b6`)
  - `CRTM_K_Matrix_Module.f90` (commit `c6da462216311df5905c554dbcb60b27685dc3b6`)
  - `CRTM_Tangent_Linear_Module.f90` (commit `a7810c441bd1191e70ee0d71beec97caab5930ee`)
  - Additionally, this was also fixed in an unspecified module (commit `640062010439ec07caa5310f7a76e747fc7d83a5`)

- **Additional updates:**
  - The allocation of `atm_K` was added in `test_Simple.f90` (commit `e80fb9fdb44a4de33ffff275d8e3c867696f3cdc`)
  - The version number was updated to `v3.0.0` in `CRTM_Version.inc` (commit `41dea3ee7e618d4e55d8d7d86be3b903105172eb`)
  - An initial commit of change was made to `CRTM_Atmosphere_Define.f90` (commit `ae06b50a214fa2f4265cb66f131dd35056e3dcc3`)

Please review these changes and merge them if everything looks good.

### Testing:

Please run the existing test suites to verify these changes. If new tests are required, please add them accordingly.

## Issue(s) addressed

Resolves #33 
Resolves #29 
Resolves #19  (hopefully)

## Checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have run the unit tests before creating the PR

### CODE REVIEW

**Comment by huishao-r** [comment]
@BenjaminTJohnson I'll check out this branch and test it.

**Comment by chengdang** [comment]
All CRTM standalone ctests passed! yay

File: src/Atmosphere/CRTM_Atmosphere_Define.f90 [hunk 1]
```diff
@@ -2462,7 +2462,7 @@ FUNCTION Read_Record( &
 
     ! RH APPROACH #2
     ! Compute the relative humidity
-    CALL Compute_Relative_Humidity( atm )
+    !CALL Compute_Relative_Humidity( atm )
```
Comment (chengdang) [inline src/Atmosphere/CRTM_Atmosphere_Define.f90:2465]: It is okay we switch to approach 1, especially for JEDI.

JEDI does not use function `Read_Record` for input profiles, but assign all data directly from geovals. For now there is no relative humidity assigned in UFO that `atm%Relative_Humidity = 0`. This should be updated later if GOCART-GEOS5 and NAAPS tables are needed for tests later.

**Review by chengdang — APPROVED** [review]

**Comment by fcvdb** [comment]
Using:
```
jedi-bundle: release/skylab-v5
ufo: feature/btj_v3.0.0_enable
crtmv3: hotfix/btj_fix_adjoint
```
On Orion with GNU, I found the following UFO tests segfaulted:
```
 194 ufo_test_tier1_test_ufo_amsr2_qc
 207 ufo_test_tier1_test_ufo_gmi_qc_filters
 208 ufo_test_tier1_test_ufo_gmi_qc_filters_geos
 210 ufo_test_tier1_test_ufo_mhs_qc_filters_geos
 295 ufo_test_tier1_test_ufo_gmi_clw_ret
 417 ufo_test_tier2_test_ufo_linopr_amsr2_crtm
 429 ufo_test_tier2_test_ufo_linopr_gmi_crtm
```
All errors occur at the same place:
```
Loading  transmittance coefficients.
Caught SIGFPE: floating-point invalid operation
 0# sigfpe_handler(int, siginfo_t*, void*) in /work2/noaa/da/fvandenb/skylab/jedi_2023062122_skylab-v5_crtm-v3_hotfix/build/bin/../lib64/liboops.so
 1# 0x00002ACF904595D0 in /lib64/libpthread.so.0
 2# __ada_module_MOD_crtm_ada in /work2/noaa/da/fvandenb/skylab/jedi_2023062122_skylab-v5_crtm-v3_hotfix/build/bin/../lib64/libcrtm.so
```

File: src/Atmosphere/CRTM_Atmosphere_Define.f90 (hunk 1 above)
Comment (BenjaminTJohnson) [reply src/Atmosphere/CRTM_Atmosphere_Define.f90:2465]: @chengdang before we come back to this, I'd like to do a complete code review, cleanup.  there's lots of "comments" left in the code that need to be removed.

The other thing that needs to be done is an assessment of openMP again.   I haven't tested with multi-thread openMP yet, I'll do that now.

**Comment by BenjaminTJohnson** [comment]
> Using:
> 
> ```
> jedi-bundle: release/skylab-v5
> ufo: feature/btj_v3.0.0_enable
> crtmv3: hotfix/btj_fix_adjoint
> ```
> 
> On Orion with GNU, I found the following UFO tests segfaulted:
> 
> ```
>  194 ufo_test_tier1_test_ufo_amsr2_qc
>  207 ufo_test_tier1_test_ufo_gmi_qc_filters
>  208 ufo_test_tier1_test_ufo_gmi_qc_filters_geos
>  210 ufo_test_tier1_test_ufo_mhs_qc_filters_geos
>  295 ufo_test_tier1_test_ufo_gmi_clw_ret
>  417 ufo_test_tier2_test_ufo_linopr_amsr2_crtm
>  429 ufo_test_tier2_test_ufo_linopr_gmi_crtm
> ```
> 
> All errors occur at the same place:
> 
> ```
> Loading  transmittance coefficients.
> Caught SIGFPE: floating-point invalid operation
>  0# sigfpe_handler(int, siginfo_t*, void*) in /work2/noaa/da/fvandenb/skylab/jedi_2023062122_skylab-v5_crtm-v3_hotfix/build/bin/../lib64/liboops.so
>  1# 0x00002ACF904595D0 in /lib64/libpthread.so.0
>  2# __ada_module_MOD_crtm_ada in /work2/noaa/da/fvandenb/skylab/jedi_2023062122_skylab-v5_crtm-v3_hotfix/build/bin/../lib64/libcrtm.so
> ```

@fcvdb  could you verify that you're using the correct branch?   (i.e., cd to the crtm directory and do a `git branch --remote`)   These are exactly the errors that this PR fixes.   If everything verifies, can you test with Intel also?    

Thanks for checking.

**Comment by huishao-r** [comment]
Using the branch, 
CRTM ctests all passed. But UFO still have some errors. Probably similar to what @fcvdb Francois has shown. I ran all the ctest under jedi-bundle so this should be a quite comprehensive list here (only tier1 was tested):  

The following test has issues to be fixed:

```
	968 - ufo_test_tier1_test_ufo_amsr2_qc (Failed)
                 `Caught SIGFPE: floating-point invalid operation`

	979 - ufo_test_tier1_test_ufo_gmi_qc_filters (Failed)
             Caught SIGFPE: floating-point invalid operation
            0# sigfpe_handler(int, siginfo_t*, void*) in /work2/noaa/da/hshao/work/JCSDA/jedi-bundle/build/bin/../lib64/liboops.so
             1# 0x00007F2E05F835D0 in /lib64/libpthread.so.0
            2# __ada_module_MOD_crtm_ada in /work2/noaa/da/hshao/work/JCSDA/jedi-bundle/build/bin/../lib64/libcrtm.so`

	980 - ufo_test_tier1_test_ufo_gmi_qc_filters_geos (Failed)
                  `Caught SIGFPE: floating-point invalid operation`

	1061 - ufo_test_tier1_test_ufo_gmi_clw_ret (Failed)
                    `Caught SIGFPE: floating-point invalid operation`

	1361 - fv3jedi_test_tier1_hofx_nomodel_abi_radii (Failed)
                    RTM_IRwaterCoeff_Load(FAILURE) : Error reading IRwaterCoeff IRwaterCoeff file Data/crtm/Nalli.IRwater.EmisCoeff.bin
```



The following ctests only require reference updates
```
        972 - ufo_test_tier1_test_ufo_amsua_qc_filters (Failed)
                  `Test "ufo/ObsFilters/testFilters" failed: Condition failed: passed == passedBenchmark @`

	973 - ufo_test_tier1_test_ufo_amsua_qc_filters_geos (Failed)
                   `Test "ufo/ObsFilters/testFilters" failed: Condition failed: passed == passedBenchmark`

	975 - ufo_test_tier1_test_ufo_atms_qc_filters (Failed)
                 `Test "ufo/ObsFilters/testFilters" failed: Condition failed: passed == passedBenchmark`

	976 - ufo_test_tier1_test_ufo_atms_n20_qc_filters_geos (Failed)
                 `Test "ufo/ObsFilters/testFilters" failed: Condition failed: passed == passedBenchmark`

	1204 - ufo_test_tier1_test_ufo_obsdiag_crtm_amsua_jacobian (Failed)
                    Test "ufo/ObsDiagnostics/testObsDiagnostics" failed: Condition failed

	1206 - ufo_test_tier1_test_ufo_obsdiag_crtm_atms_jacobian (Failed)
              Test "ufo/ObsDiagnostics/testObsDiagnostics" failed: Condition failed: rms < 100*tol

	1224 - ufo_test_tier1_test_ufo_ssmis_f17_gfs_backgroundcheck_bc (Failed)
                         Test "ufo/ObsFilters/testFilters" failed: Condition failed: passed == passedBenchmark

	1381 - fv3jedi_test_tier1_hyb-3dvar (Failed)
Test Line: 'CostJo   : Nonlinear Jo(AMSUA-NOAA19) = 5.9632399022412642e+02, nobs = 1015, Jo/n = 5.8751132041785858e-01, err = 8.8534579958897837e+00'
Ref Line : 'CostJo   : Nonlinear Jo(AMSUA-NOAA19) = 5.9785513464828637e+02, nobs = 1016, Jo/n = 5.8844009315776213e-01, err = 8.8497014000655270e+00'
[Orion-login-4:173662] *** Process received signal ***
[Orion-login-4:173662] Signal: Aborted (6)
[Orion-login-4:173662] Signal code:  (-6)
```

**Comment by huishao-r** [comment]
@BenjaminTJohnson For CRTM, it does resolve the ctest issues. I'll try to see whether it works out for the skylab just for a quick test. Then we can decide whether this goes to skylab release (CRTMv3 develop). Meanwhile, we will comment out UFO failure ctests/modify the reference for the UFO release branch.

**Comment by fcvdb** [comment]
> @fcvdb could you verify that you're using the correct branch? (i.e., cd to the crtm directory and do a `git branch --remote`) These are exactly the errors that this PR fixes. If everything verifies, can you test with Intel also?
> 
> Thanks for checking.

Ben. That may be the issue:
 I cloned the hotfix branch
```
git branch -a
  develop
* hotfix/btj_fix_adjoint
```
But that may not be enough:
 ```
git branch --remote
  origin/HEAD -> origin/develop
```
Let me try again.

**Review by BenjaminRuston — APPROVED** [review]
this works for me with skylab-v5, was having trouble with TEMPEST and G17 and now resolved

**Review by huishao-r — APPROVED** [review]
